File: go/adk/pkg/fileextract/fileextract.go
// Package fileextract turns uploaded text-like file blobs into text the model can read.
package fileextract

import (
	"fmt"
	"path/filepath"
	"strings"

	"google.golang.org/genai"
)

// textExts are treated as text when the browser sends no useful MIME type.
var textExts = map[string]bool{
	".txt": true, ".md": true, ".markdown": true, ".csv": true, ".tsv": true, ".log": true,
	".json": true, ".ndjson": true, ".yaml": true, ".yml": true, ".xml": true, ".toml": true,
	".ini": true, ".conf": true, ".sh": true, ".py": true, ".go": true, ".js": true, ".ts": true,
}

// normalizeMIME strips parameters (e.g. "; charset=utf-8") and lowercases.
func normalizeMIME(mimeType string) string {
	mimeType = strings.ToLower(strings.TrimSpace(mimeType))
	if i := strings.IndexByte(mimeType, ';'); i >= 0 {
		mimeType = strings.TrimSpace(mimeType[:i])
	}
	return mimeType
}

// isTextLikeMIME reports whether a MIME type can be inlined as raw text.
func isTextLikeMIME(mimeType string) bool {
	mimeType = normalizeMIME(mimeType)
	if strings.HasPrefix(mimeType, "text/") {
		return true
	}
	switch mimeType {
	case "application/json", "application/xml", "application/x-ndjson",
		"application/yaml", "application/x-yaml":
		return true
	}
	return false
}

// extractFileText turns a file's bytes into text, or errors for formats that
// have none (e.g. arbitrary binary).
func extractFileText(data []byte, mimeType, name string) (string, error) {
	if isTextLikeMIME(mimeType) {
		return string(data), nil
	}
	if m := normalizeMIME(mimeType); (m == "" || m == "application/octet-stream") && textExts[strings.ToLower(filepath.Ext(name))] {
		return string(data), nil
	}
	return "", fmt.Errorf("unsupported file type for text extraction: mime=%q name=%q", mimeType, name)
}

// maxTextChars caps the text taken from one file.
const maxTextChars = 200_000

// InlineFileToText converts a non-image file blob named name into chat text.
// Unsupported types become a short note so the model can say the file was unreadable.
func InlineFileToText(blob *genai.Blob, name string) string {
	if blob == nil {
		return ""
	}
	if name == "" {
		name = "file"
	}
	text, err := extractFileText(blob.Data, blob.MIMEType, name)
	if err != nil {
		return fmt.Sprintf("[Uploaded file %q (%s) could not be read as text.]", name, blob.MIMEType)
	}
	if strings.TrimSpace(text) == "" {
		return fmt.Sprintf("[Uploaded file %q (%s) contained no extractable text.]", name, blob.MIMEType)
	}
	if runes := []rune(text); len(runes) > maxTextChars {
		text = string(runes[:maxTextChars]) + "\n\n[truncated]"
	}
	return fmt.Sprintf("Contents of uploaded file %q:\n\n%s", name, text)
}

File: python/packages/kagent-adk/src/kagent/adk/models/_file_extract.py
"""Helpers to turn uploaded file blobs into text the model can read.

Rich documents (PDF, DOCX, XLSX, PPTX, HTML) are extracted to markdown via
``markitdown``; text-like files are returned as-is. This mirrors the Go ADK
runtime behaviour so non-image uploads reach the model instead of being dropped.
"""

from __future__ import annotations

import os
import tempfile
from typing import Optional

from google.genai import types

# Rich document MIME types mapped to the extension markitdown uses for format
# detection. Kept in sync with the Go runtime (fileextract.go): the common
# formats both markitdown (Python) and tabula (Go) support — PDF, DOCX, XLSX,
# PPTX, HTML, EPUB.
_DOC_MIME_TO_EXT: dict[str, str] = {
    "application/pdf": ".pdf",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document": ".docx",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": ".xlsx",
    "application/vnd.openxmlformats-officedocument.presentationml.presentation": ".pptx",
    "application/epub+zip": ".epub",
    "text/html": ".html",
}

# Filename extensions markitdown can extract as rich documents.
_DOC_EXTS: dict[str, str] = {
    ".pdf": ".pdf",
    ".docx": ".docx",
    ".xlsx": ".xlsx",
    ".pptx": ".pptx",
    ".epub": ".epub",
    ".html": ".html",
    ".htm": ".html",
}


def _normalize_mime(mime_type: Optional[str]) -> str:
    if not mime_type:
        return ""
    mime_type = mime_type.strip().lower()
    if ";" in mime_type:
        mime_type = mime_type.split(";", 1)[0].strip()
    return mime_type


def _is_text_like(mime_type: Optional[str]) -> bool:
    """Whether a MIME type can be inlined as raw text (text/html excluded)."""
    mime_type = _normalize_mime(mime_type)
    if mime_type == "text/html":
        return False
    if mime_type.startswith("text/"):
        return True
    return mime_type in {
        "application/json",
        "application/xml",
        "application/x-ndjson",
        "application/yaml",
        "application/x-yaml",
    }


def _doc_ext(mime_type: Optional[str], name: Optional[str]) -> str:
    """Extension to use for markitdown, derived from filename then MIME type."""
    if name:
        _, ext = os.path.splitext(name.lower())
        if ext in _DOC_EXTS:
            return _DOC_EXTS[ext]
    return _DOC_MIME_TO_EXT.get(_normalize_mime(mime_type), "")


def _extract_doc_text(data: bytes, ext: str) -> str:
    """Write bytes to a temp file (markitdown detects by extension) and extract."""
    # Lazy import so the package still imports if markitdown is unavailable.
    from markitdown import MarkItDown

    fd, tmp_path = tempfile.mkstemp(suffix=ext)
    try:
        with os.fdopen(fd, "wb") as f:
            f.write(data)
        result = MarkItDown().convert(tmp_path)
        return result.text_content or ""
    finally:
        os.remove(tmp_path)


def extract_file_text(data: bytes, mime_type: Optional[str], name: Optional[str]) -> str:
    """Extract readable text from an uploaded file's bytes.

    Raises ValueError for formats that cannot be represented as text.
    """
    ext = _doc_ext(mime_type, name)
    if ext:
        return _extract_doc_text(data, ext)
    if _is_text_like(mime_type):
        return data.decode("utf-8", errors="replace")
    raise ValueError(f"unsupported file type for text extraction: mime={mime_type!r} name={name!r}")


def inline_file_to_text(blob: types.Blob) -> Optional[str]:
    """Convert a non-image inline file blob into text for a chat message.

    Returns ``None`` for an empty blob. On extraction failure returns a short
    note so the model can tell the user the file could not be read, instead of
    silently dropping it.
    """
    if blob is None or not blob.data:
        return None
    name = blob.display_name or "file"
    mime_type = blob.mime_type or ""
    try:
        text = extract_file_text(blob.data, mime_type, name)
    except Exception:
        return f'[Uploaded file "{name}" ({mime_type}) could not be read as text.]'
    if not text.strip():
        return f'[Uploaded file "{name}" ({mime_type}) contained no extractable text.]'
    return f'Contents of uploaded file "{name}":\n\n{text}'
